### ISSUE TITLE
Word embedding output dimensions

### ISSUE THREAD

**Issue by stunax** [dotnet/machinelearning#2068]
### System information

- **OS version/distro**: windows 64bit 10.0.17134
- **.NET Version (eg., dotnet --info)**: 2.1.502

### Issue
I am having trouble understanding the output of the WordEmbeddings function. Using a model with 50 dimensionality output, I would expect the embedding dimensions to match that output in some way. It only just multiplies the expected output dimension by 3.

I have checked the model files, and they have the correct output for each word for each model.

- **What did you do?** 
Created wordEmbedding using several different preprocessed models. Both default and downloaded elsewhere
- **What happened?**
For GloVe50D the output dimensions were 150.
For FastTextWikipedia300D the output was 900
- **What did you expect?**
I expected the output to match the embedding size in the models.
For GloVe50D to output dimensions 50 * input size, or [inputsize][50].
For FastTextWikipedia300D to output dimensions 300 * input size, or [inputsize][300].

### Source code / logs

Please paste or attach the code or logs or traces that would be helpful to diagnose the issue you are reporting.

This code illustrates my issue. I would assume the output to be of variable size, but all outputs have dimensionality of 150, which I have trouble grasping. I have tried forcing the output to a variable vector output but it changes nothing.

What worries me most, is what happens, when the input is a single word sentence, as I can not fathom or figure out what the extra 100 values are.

```

using System;
using System.Linq;
using Microsoft.ML;
using Microsoft.ML.Runtime.Api;
using Microsoft.ML.Runtime.Data;
using Microsoft.ML.StaticPipe;
using Microsoft.ML.Transforms.Text;

namespace ConsoleApp1
{
    class DataEntry
    {
        public DataEntry(string message, int label)
        {
            Message = message;
            Label = label;
        }
        public string Message { get; set; }
        public int Label { get; set; }
    }

    class ModelOutput
    {
        public int Label;
        public float[] Embedding;
    }

    class Program
    {
        private static MLContext mLContext = new MLContext();
        private static PredictionFunction<DataEntry, ModelOutput> wordEmbedder;

        static void Main()
        {
            var dataEntries = new[]
            {
                "I am a horse",
                "cow",
                "The cat is norse",
                "Don't ask me how",
                "This is a sentence that is very long, and should have a different dimension than the other ones?"
            };
            var labels = dataEntries.Select((_, i) => i).ToArray();

            TokenizeData(dataEntries, labels, out  var data);
            
            Console.WriteLine($"{data.Length} {data[0].Length}");
            foreach (var d in data)
            {
                Console.WriteLine($"{d.Length}");
            }

            Console.Read();
        }
        

        private static void TokenizeData(string[] dataEntries, int[] labels, out double[][] dataVectors)
        {
            var dataEnum = dataEntries.Select((s, i) => new DataEntry(s, labels[i])).ToArray();
            var reader = mLContext.CreateDataView(dataEnum).AssertStatic(
                mLContext, c =>
                (
                    Message: c.Text.Scalar,
                    Label: c.I4.Scalar));

            // Inspect the message texts that are read from the file.
            var pipeline = reader.MakeNewEstimator()
                    .Append(r => (
                        LabelMid: r.Label,
                        Embedding: r.Message
                            .NormalizeText()
                            .TokenizeText()
                            .RemoveStopwords()
                            .WordEmbeddings(
                                WordEmbeddingsExtractingTransformer
                                    .PretrainedModelKind
                                    .GloVe50D)
                    ))
                ;

            var modelFit = pipeline.Fit(reader);

            wordEmbedder = modelFit.AsDynamic.MakePredictionFunction<DataEntry, ModelOutput>(mLContext);

            dataVectors = dataEntries.Select((s, i) => NameToTfVector(s, labels[i])).ToArray();
        }
        
        private static double[] SentenceFromWordEmbedding(ModelOutput embedding)
        {
            return embedding.Embedding.Select(f => (double) f).ToArray();
        }

        private static double[] NameToTfVector(string inputText, int label)
        {
            return SentenceFromWordEmbedding(wordEmbedder.Predict(new DataEntry(inputText, label)));
        }
    }
}
```
Output:
`5 150
150
150
150
150
150`

**Comment by justinormont** [comment]
Greetings @stunax! 

The WordEmbedding transform produces an output consisting of the average, min, and max of the words in the phrase. The min/max provide a bounding hyper-rectangle for the words in the word embedding space. This can assist for longer phrases where the average of many words drowns out the useful signal and your label is reasonable correlated with a dimension of the embedding space. 

For instance, if you're trying to classify ( trucks vs. dogs ), a dimension of the pretrained embedding space may reasonably represent "mechanical-ness" or "softness"; and any word in your phrase peaking these dimensions would push the min/max outwards even when the average is overwhelmed by other words.

If you want to experiment, you can try to select the first/middle/last 3rd of the output vector as your Features column and see the impact of the average/min/max on your dataset.

Let us know if you find the WordEmbedding valuable on your dataset, and if any of the pretrained custom embeddings which you're trying turn out to be useful useful.

**Comment by stunax** [comment]
Thank you very much! 

This was what I needed to understand the output. I was unable to find any documentation for the output, but this makes sense.

I ended up just loading the model manually, but glove50 improved model performance, by reducing test error by 2/3.

I am abit curious as how your FastText model works? As far as i understand, the benefit of FastText comes from the subword information, but I miss an example as how it should be combined with the ngram transformer.

**Comment by justinormont** [comment]
FastText uses the sub-words (often 3 to 6 length charactergrams) plus ngrams (often length ~2 ngrams) to train the model. When they flatten the model to a lookup table, they include only the unigrams. We use the unigrams as the key to the lookup table. 

I believe the binary format of FastText embeddings (.bin files) retains the full neural net includes inputs for the charactergrams. We haven't investigated the gains by using the binary format vs. the flattened (.vec) format. I expect this would mainly help when there are words which are telling of the label, but not in the pre-trained word embedding file, as it can lookup the chargrams even without the full word existing. 

For handling the case of out-of-vocabulary words, the trichargram portion of the TextTransform provides similar gains. Using bigrams+trichargrams as a feature gives you a pseudo-stemming ability from the trichargrams. I tend to use a word embedding in parallel with these ngrams+chargrams.

**Comment by Ivanidzo4ka** [comment]
This is clear sign what we need better documentation.
